Fixed issue with num_challenges of the BabyBear configuration

DDT92 opened the pull request:
Permutation argument and combining constraints using a single challenge from a field `F` have a soundness error  `e = degree / |F|`  (see section 3.4 of Plonky2 document).  This error is too large for practical usage, so we need `n` challenges to reduce the error to `e^n`.
While `n=2` is ok for the targeted security bits for Goldilocks, it is far from being sufficient for BabyBear: if, for example, `degree = 2^14` we need at least 6 challenges to achieve 100 bits security: `((2^14)/(2^31))^6 = 2 ^ (-102)`.
The increases in the number of challenges unfortunately causes the impossibility to achieve` 2^12` rows for the circuit of the recursive verifier, so it's convenient to use a narrow configuration as default, having a recursive verifier circuit with `2^13` rows.

DDT92 committed:
Fixed issue with num_challenges of the BabyBear configuration

1) Modified the BabyBear circuit configurations in order to have 6 challenges.
2) Fixed Poseidon2R0Gate implementation  in order to be coherent with Poseidon2Gate.
3) Used the wide configuration in the tests where more routed wires are needed.

ptagl commented on plonky2/src/gates/poseidon2_babybear.rs:
```diff
@@ -31,8 +31,8 @@ use crate::plonk::config::AlgebraicHasher;
 
-const USE_INTERNAL_PERMUTATION_GATE: bool = true;
-const USE_EXTERNAL_PERMUTATION_GATE: bool = true;
+const USE_INTERNAL_PERMUTATION_GATE: bool = false;
+const USE_EXTERNAL_PERMUTATION_GATE: bool = false;
```
Can you please add a comment describing the reason why these two flags were introduced and why they are now disabled?

Thanks!

ptagl commented on plonky2/src/gates/poseidon2_risc0_babybear.rs:
```diff
@@ -45,8 +45,16 @@ const NON_ROUTED_WIRES_PER_OP: usize =
+        }
+    }
+
+    fn num_ops(config: &CircuitConfig) -> usize {
+        let wires_per_op = ROUTED_WIRES_PER_OP + NON_ROUTED_WIRES_PER_OP;
+        (config.num_wires / wires_per_op).min(config.num_routed_wires / ROUTED_WIRES_PER_OP)
     }
```
This function can now be used to compute `num_ops` inside the `new_from_config()` constructor below (credits to @DDT92).

DDT92 committed:
Added security checks for config parameters

Implements checks to ensure sufficient security bits for permutation argument, combining constraint and evaluation.

ptagl commented on plonky2/src/gates/poseidon2_risc0_babybear.rs:
```diff
@@ -205,14 +214,14 @@ impl<F: RichField + HasExtension<D>, const D: usize, const NUM_HASH_OUT_ELTS: us
                 let input_lhs = vars.local_wires[Self::wire_input(op, i)];
                 let input_rhs = vars.local_wires[Self::wire_input(op, i + SPONGE_CAPACITY)];
-                let delta_i = vars.local_wires[self.wire_delta(op, i)];
+                let delta_i = vars.local_wires[Self::wire_delta(self.num_ops, op, i)];
```
I see that the refactoring of `Poseidon2R0BabyBearGate` made the function calls aligned with the rest of the codebase, but I also noted this "ugly" side-effect that the usage of `wire_delta()` passed from

`self.wire_delta(...)`

to

`Self.wire_delta(self.num_ops, ...)`

Are we sure that static functions are the best option here?

DDT92 committed:
Minor fixes

Added a comment motivating the introduction of  USE_INTERNAL_PERMUTATION_GATE and USE_EXTERNAL_PERMUTATION_GATE.
Re-used num_ops method in new_from_config of Poseidon2R0BabyBearGate.
